packages/medusa/src/services/idempotency-key.ts

```diff
@@ -0,0 +1,197 @@
+    super({ manager, idempotencyKeyRepository })
+
+    this.manager_ = manager
+    this.idempotencyKeyRepository_ = idempotencyKeyRepository
+  }
+
+  /**
+   * Execute the initial steps in a idempotent request.
+   * @param headerKey - potential idempotency key from header
+   * @param reqMethod - method of request
+   * @param reqParams - params of request
+   * @param reqPath - path of request
+   * @return the existing or created idempotency key
+   */
+  async initializeRequest(
+    headerKey: string,
+    reqMethod: string,
+    reqParams: Record<string, unknown>,
+    reqPath: string
+  ): Promise<IdempotencyKey> {
+    return await this.atomicPhase_(async () => {
+      return (
+        (await this.retrieve(headerKey).catch(() => void 0)) ??
+        (await this.create({
+          request_method: reqMethod,
+          request_params: reqParams,
+          request_path: reqPath,
+        }))
+      )
+    }, "SERIALIZABLE")
+  }
+
+  /**
+   * Creates an idempotency key for a request.
+   * If no idempotency key is provided in request, we will create a unique
+   * identifier.
+   * @param payload - payload of request to create idempotency key for
+   * @return the created idempotency key
+   */
+  async create(payload: CreateIdempotencyKeyInput): Promise<IdempotencyKey> {
+    return await this.atomicPhase_(async (manager) => {
+      const idempotencyKeyRepo = manager.getCustomRepository(
+        this.idempotencyKeyRepository_
+      )
+
+      payload.idempotency_key = payload.idempotency_key ?? v4()
+
+      const created = await idempotencyKeyRepo.create(payload)
+      return await idempotencyKeyRepo.save(created)
+    })
+  }
+
+  /**
+   * Retrieves an idempotency key
+   * @param idempotencyKey - key to retrieve
+   * @return idempotency key
+   */
+  async retrieve(idempotencyKey: string): Promise<IdempotencyKey | never> {
+    const idempotencyKeyRepo = this.manager_.getCustomRepository(
+      this.idempotencyKeyRepository_
+    )
+
+    const iKey = await idempotencyKeyRepo.findOne({
+      where: { idempotency_key: idempotencyKey },
+    })
+
+    if (!iKey) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Idempotency key ${idempotencyKey} was not found`
+      )
+    }
+
+    return iKey
+  }
+
+  /**
+   * Locks an idempotency.
+   * @param idempotencyKey - key to lock
+   * @return result of the update operation
+   */
+  async lock(idempotencyKey: string): Promise<IdempotencyKey | never> {
+    return await this.atomicPhase_(async (manager) => {
+      const idempotencyKeyRepo = manager.getCustomRepository(
+        this.idempotencyKeyRepository_
+      )
+
+      const key = await this.retrieve(idempotencyKey)
+
+      const isLocked =
+        key.locked_at &&
+        new Date(key.locked_at).getTime() > Date.now() - KEY_LOCKED_TIMEOUT
+
+      if (isLocked) {
+        throw new MedusaError("conflict", "Key already locked")
```

Comment:
suggestion: Create a new error type `CONFLICT` similar to how we have `NOT_FOUND` etc.

Reply:
👌